### PR TITLE
[CodeParrot] Near-deduplication with jaccard similarity

### PR DESCRIPTION
**Pull request by liyongsea**
# What does this PR do?

This PR address the code duplication issue describe in this thread
https://twitter.com/miltos1/status/1497126435261083649?s=20&t=v5-vwaEtXLrgZ_GuZHrPKQ

# run the code

```
from datasets import load_dataset
from minhash_deduplication import deduplicate_dataset
ds = load_dataset("lvwerra/codeparrot-clean", split="train")
ds_dedup, duplicate_clusters = deduplicate_dataset(ds)
``` 

The function runs in 2:30 (make_duplicate_clusters) + 1:30 (find_extremes) on a 8 cores VM
```
Orginal dataset size: 5361373
Duplicate cluster: 757944
Files in duplicate cluster: 2677040
Unique files in duplicate cluster: 911947
Filtered dataset size: 3596280
```

### CODE REVIEW

**Review by lvwerra — COMMENTED** [review]
Hi @liyongsea 

Thanks for adding deduplication, that's an exciting addition! In terms of structure, I think it would be good to include this in the main `preprocessing.py`. Since it requires quite a bit of code we could probably leave most of the code in `minhash_deduplication.py` and then do something like the following in `preprocessing.py`:

```Python
from minhash_deduplication import deduplicate_dataset

# other preprocessing steps

ds = deduplicate_dataset(ds, arg1, arg2, ...)

# save dataset and push to hub

```

I like that you'll use `dataset.map()` for the paralellization - it matches well with the rest of the codebase. You can probably also do `minhash_iter` with a simple `map`.

**Comment by liyongsea** [comment]
Hi @lvwerra I agree with you. I will do that. The overall code is running now, here are the next steps:

- [ ] refactor the code to be used in preprocess.py and clean up
- [ ] document statistics and performance data in the PR
- [ ] use dataset.map to compute minhash

I will do the deduplication of the validation set in another PR probably.
question, does dataset.map put the whole dataset in RAM? I imagine it is not a problem because preprocess.py is already doing so

**Comment by liyongsea** [comment]
Hi @lvwerra I agree with you. I will do that. The overall code is running now, here are the next steps:

- [ ] refactor the code to be used in preprocess.py and clean up
- [ ] document statistics and performance data in the PR
- [ ] use dataset.map to compute minhash

I will do the deduplication of the validation set in another PR probably.
question, does dataset.map put the whole dataset in RAM? I imagine it is not a problem because preprocess.py is already doing so

**Comment by liyongsea** [comment]
Hi @lvwerra there is one decision we need make, then the PR will be ready to review.
I mentioned before, we could use dataset.map to compute the minhash. However, there is two steps in the deduplication:
- compute minhash for each code file
- add into MinHashLSH (can not be parallelized)

In previous function, a queue is used while adding into minhash. It would be difficult to do the same using dataset.map. So the dataset.map implementation will be almost twice slow (to be confirmed ...) 
I might prefer the dataset.map solution, which makes the code easier to read

**Comment by liyongsea** [comment]
Here are some statistic and time performance data:

on the dataset lvwerra/codeparrot-clean
Execution time 13h: Execution time: 2:30:00 for make_duplicate_clusters, 11:00:00 for find_cluster_extremes

Orginal dataset size: 5361373                                                                        
Duplicate cluster: 757938          
Files in duplicate cluster: 2677039          
Unique files in duplicate cluster: 940857
Filtered dataset size: 3625191 

I think the code is ready for review. If you need to generate a dataset, you can go ahead. I might still need more days to figure out how to do find_cluster_extremes better

**Comment by liyongsea** [comment]
multipro_find_extremes is done with multi processing ! This PR is ready for review
Execution time ~3h: Execution time: 2:30:00 for make_duplicate_clusters, 1:00:00 for multipro_find_extremes

Orginal dataset size: 5361373
Duplicate cluster: 757938
Files in duplicate cluster: 2677039
Unique files in duplicate cluster: 940857
Filtered dataset size: 3625191
@lvwerra when review, pay more attention to
- [Here](https://github.com/huggingface/transformers/pull/17054/files#diff-c6c3b9ed9e98c7b3b8603011208c2db7c0cf08facbf2a75ec8f56dfea0242040R140) I use a global parameter to be able to do multi pro in a efficient way

File: examples/research_projects/codeparrot/scripts/minhash_deduplication.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+    def add(self, filename: str, min_hash: MinHash) -> None:
+        close_duplicates = self.__index.query(min_hash)
+        if filename in self.__index.keys:
+            print("Duplicate key %s" % filename)
+            return
+
+        self.__index.insert(filename, min_hash)
+        if len(close_duplicates) > 0:
+            # print("`%s` duplicate of: %s" % (filename, close_duplicates))
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/minhash_deduplication.py:58]:
```suggestion
```

File: examples/research_projects/codeparrot/scripts/minhash_deduplication.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+            if jaccard_similarity(code1, code2) >= 0.85:
+                element2["copies"] += 1
+                break
+        else:
+            element1["copies"] = 1
+            extremes.append(element1)
+    return extremes
+
+
+def multipro_find_extremes(cluster_list, dataset):
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/minhash_deduplication.py:139]: I would name the function by what it does not how it works:
```suggestion
def find_extremes(cluster_list, dataset):
```

File: examples/research_projects/codeparrot/scripts/minhash_deduplication.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        Each cluster is a list of dicts with the following keys:
+        - base_index : int
+            The index of the code in the original dataset.
+        - repo_name : str
+        - path : str
+        - copies : int
+            The number of copies of the code in the cluster. (find_cluster_extremes)
+        - is_extreme : bool
+            Whether the code is an extreme in the cluster.
+        All the codes in the cluster are removed from the dataset except the extremes.
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/minhash_deduplication.py:181]: Could you follow [this](https://github.com/huggingface/transformers/blob/71d18d08319db50a1a5bd022aed54d2613eec3b9/src/transformers/pipelines/base.py#L182-L196) convention for the docstring with params/returns/types?

File: examples/research_projects/codeparrot/scripts/minhash_deduplication.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    """
+    duplicate_clusters = make_duplicate_clusters(dataset)
+    duplicate_indices = set(x["base_index"] for cluster in duplicate_clusters for x in cluster)
+    # todo: make this multiprocessing, pay attention to memory usage
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/minhash_deduplication.py:185]: Is this `TODO` done with the new function below?

File: examples/research_projects/codeparrot/scripts/minhash_deduplication.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+    ds_filter = dataset.filter(lambda x, idx: idx not in remove_indices, with_indices=True)
+
+    # update duplicate_clusters
+    for cluster in duplicate_clusters:
+        for element in cluster:
+            element["is_extreme"] = element["base_index"] in extreme_dict
+            if element["is_extreme"]:
+                element["copies"] = extreme_dict[element["base_index"]]["copies"]
+
+    print("Orginal dataset size: %d" % len(dataset))
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/minhash_deduplication.py:201]: nit: I have a preference to use f-strings in these cases:
```suggestion
    print(f"Orginal dataset size: {len(dataset)}")
```

File: examples/research_projects/codeparrot/scripts/preprocessing.py [hunk 6]
```diff
@@ -79,45 +82,58 @@ def filter(example, uniques, args):
...
+    print(f"Size of deduplicate dataset: {len(ds_dedup)}")
+
+    # Save data in batches of samples_per_file
+    output_dir = Path(args.output_dir)
+    output_dir.mkdir(exist_ok=True)
+
+    # save duplicate_clusters in the output_dir as artifacts
+    # not sure it is the right place the save it
+    with open(output_dir / "duplicate_clusters.json", "w") as f:
+        json.dump(duplicate_clusters, f)
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/preprocessing.py:131]: We will push the `output_dir` to the hub and load it as a dataset later. So we don't want the that to be in the same folder as the data files. A solution is to store the data files in `output_dir / data / f"file-{file_number+1:012}.json"` and keep the duplicate clusters where they are. I think then `datasets` will only load the files in `data/`.

File: examples/research_projects/codeparrot/scripts/preprocessing.py [hunk 7]
```diff
@@ -79,45 +82,58 @@ def filter(example, uniques, args):
...
-if not os.path.exists(args.output_dir):
-    os.makedirs(args.output_dir)
-t_start = time.time()
-for file_number, index in enumerate(range(0, len(ds_filter), args.samples_per_file)):
-    file_path = f"{args.output_dir}/file-{file_number+1:012}.json"
-    end_index = min(len(ds_filter), index + args.samples_per_file)
-    ds_filter.select(list(range(index, end_index))).to_json(file_path)
-    compress_file(file_path)
-print(f"Time to save dataset: {time.time()-t_start:.2f}")
+if __name__ == "__main__":
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/preprocessing.py:90]: I guess this is necessary because of multi processing?

**Review by lvwerra — COMMENTED** [review]
A few comments:

- We should add info about the new method and how to use it to the `README.md`
- Could you add a minimal docstring to the functions?

File: examples/research_projects/codeparrot/scripts/test_deduplicate.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
+from unittest import TestCase
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/test_deduplicate.py:1]: This should go into a folder `scripts/tests/`. Could you also add the command to run the test to the README?

File: examples/research_projects/codeparrot/scripts/preprocessing.py [hunk 9]
```diff
@@ -79,45 +82,58 @@ def filter(example, uniques, args):
...
+
+    # Deduplicate data and apply heuristics
+    t_start = time.time()
+    ds_filter = ds.filter(filter, fn_kwargs={"uniques": uniques, "args": args})
+    print(f"Time to filter dataset: {time.time()-t_start:.2f}")
+    print(f"Size of filtered dataset: {len(ds_filter)}")
+
+    # Deduplicate with minhash and jaccard similarity
+    t_start = time.time()
+    ds_dedup, duplicate_clusters = deduplicate_dataset(ds_filter)
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/preprocessing.py:120]: I think at least the Jaccard threshold is something that we want to add as a parameter here, so we can experiment with the value. Could you add it to the argument parser as well?

File: examples/research_projects/codeparrot/scripts/minhash_deduplication.py [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+                    break
+            else:
+                self.__duplicate_clusters[close_duplicates[0]].add(filename)
+
+    def get_duplicate_clusters(self) -> List[List[Dict]]:
+        duplicate_clusters = []
+        for base, duplicates in self.__duplicate_clusters.items():
+            cluster = [base] + list(duplicates)
+            # reformat the cluster to be a list of dict
+            cluster = [{"base_index": el[0], "repo_name": el[1], "path": el[2]} for el in cluster]
```
Comment (loubnabnl) [inline examples/research_projects/codeparrot/scripts/minhash_deduplication.py:72]: Maybe we can also keep the `size` information if it doesn't slow the process, could be interesting to see how large are the duplicated files and how much data we remove (ofc the latter can also be computed after the filtering)

File: examples/research_projects/codeparrot/scripts/minhash_deduplication.py [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+
+    # update duplicate_clusters
+    for cluster in duplicate_clusters:
+        for element in cluster:
+            element["is_extreme"] = element["base_index"] in extreme_dict
+            if element["is_extreme"]:
+                element["copies"] = extreme_dict[element["base_index"]]["copies"]
+
+    print("Orginal dataset size: %d" % len(dataset))
+    print("Duplicate cluster: %d" % len(duplicate_clusters))
```
Comment (loubnabnl) [inline examples/research_projects/codeparrot/scripts/minhash_deduplication.py:202]:
```suggestion
    print("Number of duplicate clusters: %d" % len(duplicate_clusters))
```

File: examples/research_projects/codeparrot/scripts/preprocessing.py [hunk 12]
```diff
@@ -79,45 +82,58 @@ def filter(example, uniques, args):
...
+    t_start = time.time()
+    ds = load_dataset(args.dataset_name, split="train")
+    print(f"Time to load dataset: {time.time()-t_start:.2f}")
+
+    # Run preprocessing
+    t_start = time.time()
+    ds = ds.map(preprocess, num_proc=args.num_workers)
+    print(f"Time to preprocess dataset: {time.time()-t_start:.2f}")
+
+    # Deduplicate hashes
```
Comment (loubnabnl) [inline examples/research_projects/codeparrot/scripts/preprocessing.py:107]: For now we're running the script on `codeparrot-clean` where there are no exact duplicates, but maybe we can still add an argument to choose if we want to do exact deduplication or near deduplication? @lvwerra

File: examples/research_projects/codeparrot/scripts/minhash_deduplication.py (hunk 5 above)
Comment (loubnabnl) [inline examples/research_projects/codeparrot/scripts/minhash_deduplication.py:201]:
```suggestion
    print(f"Original dataset size: {len(dataset)}")
```

File: examples/research_projects/codeparrot/scripts/minhash_deduplication.py [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+    return set([t for t in NON_ALPHA.split(code) if len(t.strip()) > 0])
+
+
+class DuplicationIndex:
+    def __init__(
+        self,
+        *,
+        duplication_jaccard_threshold: float = 0.85,
+        num_perm: int = 256,
+        min_num_tokens: int = 10,
```
[the remark this answers is not in the document]
Comment (liyongsea) [reply examples/research_projects/codeparrot/scripts/minhash_deduplication.py:41]: will do !

File: examples/research_projects/codeparrot/scripts/preprocessing.py (hunk 7 above)
Comment (liyongsea) [reply examples/research_projects/codeparrot/scripts/preprocessing.py:90]: i think it is good practice. I understand why you want to remove it. Let's keep it like before

File: examples/research_projects/codeparrot/scripts/minhash_deduplication.py [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+    ):
+        self.__duplication_jaccard_threshold = duplication_jaccard_threshold
```
[the remark this answers is not in the document]
Comment (liyongsea) [reply examples/research_projects/codeparrot/scripts/minhash_deduplication.py:43]: of course

File: examples/research_projects/codeparrot/scripts/minhash_deduplication.py (hunk 4 above)
Comment (liyongsea) [reply examples/research_projects/codeparrot/scripts/minhash_deduplication.py]: exactly !

File: examples/research_projects/codeparrot/scripts/minhash_deduplication.py (hunk 3 above)
Comment (liyongsea) [reply examples/research_projects/codeparrot/scripts/minhash_deduplication.py:178]: will do !

File: examples/research_projects/codeparrot/scripts/test_deduplicate.py (hunk 8 above)
Comment (liyongsea) [reply examples/research_projects/codeparrot/scripts/test_deduplicate.py]: yes

File: examples/research_projects/codeparrot/scripts/minhash_deduplication.py (hunk 10 above)
Comment (liyongsea) [reply examples/research_projects/codeparrot/scripts/minhash_deduplication.py:70]: Do you mean the size of the piece of code?

File: examples/research_projects/codeparrot/scripts/preprocessing.py (hunk 9 above)
Comment (liyongsea) [reply examples/research_projects/codeparrot/scripts/preprocessing.py]: Done !

File: examples/research_projects/codeparrot/scripts/minhash_deduplication.py (hunk 10 above)
Comment (loubnabnl) [reply examples/research_projects/codeparrot/scripts/minhash_deduplication.py:70]: yes I guess we have it in the original data

File: examples/research_projects/codeparrot/README.md [hunk 15]
```diff
@@ -49,6 +49,7 @@ The raw dataset contains many duplicates. We deduplicated and filtered the datas
 - filtering with a probability of 0.7  of files without a mention of the keywords `def` , `for`, `while`  and `class`
 - filtering files that use the assignment operator `=` less than 5 times 
 - filtering files with ratio between number of characters and number of tokens after tokenization < 1.5 (the average ratio is 3.6)
+- deduplication using minhash and jaccard similarity. minhash with a jaccard threshold 0.85 is first used to create duplicate clusters. Then these clusters are then reduced to unique files based on the exact jaccard similarity. See minhash_deduplication.deduplicate_dataset for detailed description.
```
Comment (loubnabnl) [inline examples/research_projects/codeparrot/README.md:52]:
```suggestion
- near deduplication using Minhash and Jaccard similarity. Minhash with a Jaccard threshold 0.85 is first used to create duplicate clusters. Then these clusters are then reduced to unique files based on the exact Jaccard similarity. See `minhash_deduplication.deduplicate_dataset` for detailed description.
```

Comment (loubnabnl) [reply examples/research_projects/codeparrot/README.md:52]: Maybe you can move this comment up right after the line about exact deduplication

**Review by loubnabnl — COMMENTED** [review]
Thanks Jia, the PR looks good just a small comment about the Readme

File: examples/research_projects/codeparrot/scripts/preprocessing.py (hunk 6 above)
Comment (liyongsea) [reply examples/research_projects/codeparrot/scripts/preprocessing.py]: done

**Comment by liyongsea** [comment]
We are good to go, welcome your thought @lvwerra 
I will try to run some last test

File: examples/research_projects/codeparrot/README.md [hunk 16]
```diff
@@ -40,6 +40,7 @@ The source of the dataset is the GitHub dump available on Google's [BigQuery](ht
 The raw dataset contains many duplicates. We deduplicated and filtered the dataset using the heuristics proposed in OpenAI's Codex [paper](https://arxiv.org/abs/2107.03374) and some new ones:
 
 - exact deduplication using each file's hash
+- near deduplication using MinHash and Jaccard similarity. MinHash with a Jaccard threshold 0.85 is first used to create duplicate clusters. Then these clusters are then reduced to unique files based on the exact Jaccard similarity. See `minhash_deduplication.deduplicate_dataset` for detailed description.
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/README.md:43]:
```suggestion
- near deduplication using MinHash and Jaccard similarity. MinHash with a Jaccard threshold (default=0.85) is first used to create duplicate clusters. Then these clusters are then reduced to unique files based on the exact Jaccard similarity. See `deduplicate_dataset` in `minhash_deduplication.py` for a detailed description.
```

File: examples/research_projects/codeparrot/scripts/minhash_deduplication.py [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+        for element in cluster:
+            element["is_extreme"] = element["base_index"] in extreme_dict
+            if element["is_extreme"]:
+                element["copies"] = extreme_dict[element["base_index"]]["copies"]
+
+    print(f"Original dataset size: {len(dataset)}")
+    print(f"Number of duplicate clusters: {len(duplicate_clusters)}")
+    print("Files in duplicate cluster: %d" % len(duplicate_indices))
+    print("Unique files in duplicate cluster: %d" % len(extreme_dict))
+    print("Filtered dataset size: %d" % len(ds_filter))
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/minhash_deduplication.py:268]: Can we make them all f-strings? :)

**Review by lvwerra — APPROVED** [review]
Looks good to me - just a few minor comments.

**Pull request merged**